File: go/memory/objects.go
package main

import (
	"runtime"
	"runtime/debug"

	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/common"
	"github.com/mohamedmenasy/go-vs-rust-benchmarks/go/internal/harness"
)

// obj is a 64-byte, pointer-free record (the GC never scans its contents,
// but must trace every pointer that refers to it).
type obj struct{ a, b, c, d, e, f, g, h uint64 }

const objSize = 64

func fillObj(o *obj, x uint64) {
	o.a, o.b, o.c, o.d = x, x+1, x+2, x+3
	o.e, o.f, o.g, o.h = x+4, x+5, x+6, x+7
}

// release drops all garbage and asks the runtime to return memory to the OS,
// so the harness can observe how much RSS is actually given back.
func release() {
	runtime.GC()
	debug.FreeOSMemory()
}

// ---------------------------------------------------------------- bytes

// bytesW allocates one contiguous buffer of `bytes`, fills it with
// SplitMix64 words and folds it. Go's make() zeroes memory by language
// definition; the Rust version fills a fresh Vec without zeroing.
type bytesW struct {
	words int
	seed  uint64
}

func (w *bytesW) Run() uint64 {
	buf := make([]uint64, w.words)
	r := common.NewSplitMix64(w.seed)
	for i := range buf {
		buf[i] = r.Next()
	}
	var x uint64
	for _, v := range buf {
		x ^= v
	}
	return x
}

func (w *bytesW) Teardown() { release() }

func (w *bytesW) Extra() map[string]any {
	return map[string]any{"payload_bytes": w.words * 8}
}

func setupBytes(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("bytes", 100<<20))
	return &bytesW{words: n / 8, seed: uint64(p.Int("seed", 5))}, harness.Work{Unit: "bytes", PerRun: float64(n / 8 * 8)}, nil
}

// ---------------------------------------------------------------- objects-boxed

// boxedW allocates bytes/64 objects individually on the heap, keeps them in
// a slice of pointers, reads them back, and lets them die.
type boxedW struct {
	n    int
	seed uint64
}

func (w *boxedW) Run() uint64 {
	ptrs := make([]*obj, w.n)
	r := common.NewSplitMix64(w.seed)
	for i := range ptrs {
		o := &obj{}
		fillObj(o, r.Next())
		ptrs[i] = o
	}
	var acc uint64
	for _, o := range ptrs {
		acc += o.a ^ o.h
	}
	return acc
}

func (w *boxedW) Teardown() { release() }

func (w *boxedW) Extra() map[string]any {
	return map[string]any{"objects": w.n, "object_bytes": objSize, "payload_bytes": w.n * objSize}
}

func setupBoxed(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("bytes", 100<<20)) / objSize
	return &boxedW{n: n, seed: uint64(p.Int("seed", 6))}, harness.Work{Unit: "objects", PerRun: float64(n)}, nil
}

// ---------------------------------------------------------------- objects-inline

// inlineW is the same data as boxedW stored by value in one slice.
type inlineW struct {
	n    int
	seed uint64
}

func (w *inlineW) Run() uint64 {
	objs := make([]obj, w.n)
	r := common.NewSplitMix64(w.seed)
	for i := range objs {
		fillObj(&objs[i], r.Next())
	}
	var acc uint64
	for i := range objs {
		acc += objs[i].a ^ objs[i].h
	}
	return acc
}

func (w *inlineW) Teardown() { release() }

func (w *inlineW) Extra() map[string]any {
	return map[string]any{"objects": w.n, "object_bytes": objSize, "payload_bytes": w.n * objSize}
}

func setupInline(p harness.Params) (harness.Instance, harness.Work, error) {
	n := int(p.Int("bytes", 100<<20)) / objSize
	return &inlineW{n: n, seed: uint64(p.Int("seed", 6))}, harness.Work{Unit: "objects", PerRun: float64(n)}, nil
}

File: rust/memory/src/objects.rs
//! `bytes`, `objects-boxed` and `objects-inline` (mirror of go/memory/objects.go).

use bench_common::serde_json::{Value, json};
use bench_common::{Instance, Params, SplitMix64, Work};

use crate::{OBJ_SIZE, Obj, release};

/// One contiguous buffer of `bytes`, filled with SplitMix64 words and folded.
/// The Vec is built without zero-initialisation (Go's make() zeroes by
/// language definition); it is freed at the end of `run`.
struct Bytes {
    words: usize,
    seed: u64,
}

impl Instance for Bytes {
    fn run(&mut self) -> u64 {
        let mut r = SplitMix64::new(self.seed);
        let buf: Vec<u64> = (0..self.words).map(|_| r.next_u64()).collect();
        let x = buf.iter().fold(0u64, |acc, &v| acc ^ v);
        drop(buf);
        x
    }
    fn has_teardown(&self) -> bool {
        true
    }
    fn teardown(&mut self) {
        release();
    }
    fn extra(&mut self) -> Option<Value> {
        Some(json!({"payload_bytes": self.words * 8}))
    }
}

pub fn setup_bytes(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("bytes", 100 << 20) as usize;
    Ok((
        Box::new(Bytes { words: n / 8, seed: p.int("seed", 5) as u64 }),
        Work { unit: "bytes", per_run: (n / 8 * 8) as f64, input_bytes: 0 },
    ))
}

/// bytes/64 objects, each individually heap-allocated (`Box`), kept in a Vec
/// of pointers, read back and dropped.
struct Boxed {
    n: usize,
    seed: u64,
}

impl Instance for Boxed {
    fn run(&mut self) -> u64 {
        let mut r = SplitMix64::new(self.seed);
        let mut ptrs: Vec<Box<Obj>> = Vec::with_capacity(self.n);
        for _ in 0..self.n {
            ptrs.push(Box::new(Obj::filled(r.next_u64())));
        }
        let acc = ptrs.iter().fold(0u64, |acc, o| acc.wrapping_add(o.a ^ o.h));
        drop(ptrs);
        acc
    }
    fn has_teardown(&self) -> bool {
        true
    }
    fn teardown(&mut self) {
        release();
    }
    fn extra(&mut self) -> Option<Value> {
        Some(json!({"objects": self.n, "object_bytes": OBJ_SIZE, "payload_bytes": self.n * OBJ_SIZE}))
    }
}

pub fn setup_boxed(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("bytes", 100 << 20) as usize / OBJ_SIZE;
    Ok((
        Box::new(Boxed { n, seed: p.int("seed", 6) as u64 }),
        Work { unit: "objects", per_run: n as f64, input_bytes: 0 },
    ))
}

/// The same objects stored by value in one Vec.
struct Inline {
    n: usize,
    seed: u64,
}

impl Instance for Inline {
    fn run(&mut self) -> u64 {
        let mut r = SplitMix64::new(self.seed);
        let objs: Vec<Obj> = (0..self.n).map(|_| Obj::filled(r.next_u64())).collect();
        let acc = objs.iter().fold(0u64, |acc, o| acc.wrapping_add(o.a ^ o.h));
        drop(objs);
        acc
    }
    fn has_teardown(&self) -> bool {
        true
    }
    fn teardown(&mut self) {
        release();
    }
    fn extra(&mut self) -> Option<Value> {
        Some(json!({"objects": self.n, "object_bytes": OBJ_SIZE, "payload_bytes": self.n * OBJ_SIZE}))
    }
}

pub fn setup_inline(p: &Params) -> Result<(Box<dyn Instance>, Work), String> {
    let n = p.int("bytes", 100 << 20) as usize / OBJ_SIZE;
    Ok((
        Box::new(Inline { n, seed: p.int("seed", 6) as u64 }),
        Work { unit: "objects", per_run: n as f64, input_bytes: 0 },
    ))
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::testutil::{golden_memory, params, run_twice};
    use bench_common::golden_hex;

    #[test]
    fn matches_golden() {
        let g = golden_memory();
        let b = &g["bytes"];
        let words = b["words"].as_u64().unwrap();
        let (mut inst, _) =
            setup_bytes(&params(&[("bytes", (words * 8).to_string()), ("seed", b["seed"].to_string())]))
                .unwrap();
        assert_eq!(run_twice(&mut inst), golden_hex(&b["run"]));
        let o = &g["objects"];
        let bytes = (o["n"].as_u64().unwrap() as usize * OBJ_SIZE).to_string();
        for setup in [setup_boxed as bench_common::harness::SetupFn, setup_inline] {
            let (mut inst, _) =
                setup(&params(&[("bytes", bytes.clone()), ("seed", o["seed"].to_string())])).unwrap();
            assert_eq!(run_twice(&mut inst), golden_hex(&o["run"]));
        }
    }
}
